### PR TITLE
adding activations guide

### PR DESCRIPTION
**Pull request by sad-**
Adding new guide on activation functions

### CODE REVIEW

**Comment by sad-** [comment]
originally reviewed by @thomelane here: https://github.com/sad-/mxnet-tutorials/pull/1

File: python/guide/module/gluon/activation.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
+
```
Comment (vishaalkapoor) [inline python/guide/module/gluon/activation.md]: 1. I would store the images in images/* 
2. Naming scheme for ping files. Perhaps prefix with activation-
3. Are SVGs possible? Theyd' scale better?

Comment (thomelane) [reply python/guide/module/gluon/activation.md]: RE #1, I spoke with Aaron and we're going for a folder for each tutorial, inside of which is an `images` folder, and an `md` file of the same name as the folder.

Comment (thomelane) [reply python/guide/module/gluon/activation.md]: So this avoids issue in #2.

File: python/guide/module/gluon/activation.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+## Activation Functions
+
+Deep neural networks are a way to express a nonlinear function with lots of parameters from input data to outputs. The nonlinearities that allow neural networks to capture complex patterns in data are referred to as activation functions. Over the course of the development of neural networks, several nonlinear activation functions have been introduced to make gradient-based deep learning tractable. 
+
+If you are looking to answer the question, 'which activation function should I use for my neural network model?', you should probably go with *ReLU*. Unless you're trying to implement something like a gating mechanism, like in LSTMs or GRU cells, then you should opt for sigmoid and/or tanh in those cells. However, if you have a working model architecture and you're trying to improve it's performance by swapping out activation functions or treating the activation function as a hyperparameter, then you may want to try hand-designed activations like SELU or a function discovered by reinforcement learning and exhaustive search like Swish. This guide describes these activation functions and others implemented in MXNet in detail.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: improve it's -> improve its

File: python/guide/module/gluon/activation.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    plt.plot(x.asnumpy(), x.grad.asnumpy())
+    plt.legend(["{} activation".format(activation), "{} gradient".format(activation)])
+```
+
+## Sigmoids
+
+### Sigmoid
+The sigmoid activation function, also known as the logistic function or logit function, is perhaps the most widely known activation owing to its [long history](https://web.stanford.edu/class/psych209a/ReadingsByDate/02_06/PDPVolIChapter8.pdf) in neural network training and appearance in logistic regression and kernel methods for classification. 
+
+The sigmoid activation is a non-linear function that transforms any real valued input to a value between 0 and 1, giving it a natural probabilistic interpration. The sigmoid takes the form of the function below
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: interpration -> interpretation

File: python/guide/module/gluon/activation.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+$$ \sigma(x) = \dfrac{1}{1 + e^x} $$ or alternatively
+
+$$ \sigma(x) = \dfrac{e^x}{e^x + 1} $$
+
+Warning: the term sigmoid is overloaded and can be used to refer to the class of 's' shaped functions or particularly to the logistic function that we've just described. In MxNet the sigmoid activation specifically refers to logistic function sigmoid.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: MxNet -> MXNet

File: python/guide/module/gluon/activation.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+$$ tanh(x) = \dfrac{e^x - e^{-x}}{e^x + e^{-x}}$$ 
+
+tanh addresses the issues of not being zero centered associated with the sigmoid activation function but still retains the vanishing gradient problems due to the gradient being asymptotically zero for values outside a narrow range of inputs. 
+
+In fact, the tanh can be rewritten as,
+
+$$tanh(x) = \dfrac{e^{2x} - 1}{e^{2x} + 1}$$
+
+which shows it's direct relation to sigmoid by the following equation:
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

File: python/guide/module/gluon/activation.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+
+![png](img/output_8_0.png)
+
+
+The use of tanh as activation functions in place of the logistic function was popularized by the success of the [LeNet architecture](http://yann.lecun.com/exdb/publis/pdf/lecun-01a.pdf) and the [methods paper](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf) by LeCun et al
+
+### SoftSign
+
+The softsign activation is an alternative to tanh that is also centered at zero but converges asymptotically to -1 and 1 polynomially instead of exponentially. This means that the softsign activation does not saturate as quickly as tanh. As such, there are a greater range of input values for which the softsign assigns an output of strictly between -1 and 1.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: SoftSign or Softsign? If SoftSign, should probably change "softsign" to "SoftSign".

File: python/guide/module/gluon/activation.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+## Rectifiers
+
+### ReLU
+ReLU, or Rectified Linear Unit is the most common activation function in convolutional neural networks and introduces a simple nonlinearity. When the value of the input into ReLU is positive, then it retains the same value. When the value is negative then it becomes zero. In equation form, the ReLU function is given as:
+
+$$ ReLU(x) = \mathtt{max}(0, x) $$
+
+ReLU was introduced to neural networks in the [paper by Hahnloser et al](https://papers.nips.cc/paper/1793-permitted-and-forbidden-sets-in-symmetric-threshold-linear-networks.pdf) and gained widespread popularity after it was shown in the [paper](https://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks.pdf) by Alex Krizhevsky et al to perform much better than sigmoid and tanh. This paper also introduced the AlexNet CNN that won the ILSVRC challenge in 2012.
+
+ReLU is the most widely used activation due to it's simplicity and performance across multiple datasets and although there have been efforts to introduce activation functions, many of them described in this tutorial, that improve on ReLU, they have not gained as much widespread adoption
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

File: python/guide/module/gluon/activation.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+```python
+visualize_activation(data, gluon.nn.Activation('relu'), "ReLU")
+```
+
+
+![png](img/output_15_0.png)
+
+
+As shown above, the ReLU activation mitigates the vanishing gradient problem associated with the sigmoid family of activations, by having a larger (infinite) range of values where it's gradient is non-zero. However, one drawback of ReLU as an activation function is a phenomenon referred to as the 'Dying ReLU', where gradient-based parameter updates can happen in such a way that the gradient flowing through a ReLU unit is always zero and the connection is never activated. This can largely be addressed by ensuring that the tuning the learning rate to ensure that it's not set too large when training ReLU networks.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

Comment (thomelane) [reply python/guide/module/gluon/activation.md]: it's at end is fine though

File: python/guide/module/gluon/activation.md (hunk 6 above)
Comment (thomelane) [reply python/guide/module/gluon/activation.md]: Should go SoftSign, because you use a lot of capitalization later on for these functions.

File: python/guide/module/gluon/activation.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+### SoftReLU
+
+SoftReLU also known as SmoothReLU or SoftPlus is a nonlinear activation function that takes the form
+
+$$ SoftReLU(x) = log(1 + e^x)$$
+
+The SoftReLU can be seen as a smooth version of the ReLU by observing that it's derivative is the sigmoid, seen below, which is a smooth version of the gradient of the ReLU shown above.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

File: python/guide/module/gluon/activation.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+
+![png](img/output_23_0.png)
+
+
+The activation function and activation gradient of PReLU have the same shape as LeakyRELU
+
+### ELU
+
+The ELU or exponential linear unit introduced by [Clevert et al](https://arxiv.org/abs/1511.07289) also addresses the vanishing gradient problem like ReLU and it's variants but unlike the ReLU family, ELU allows negative values which may allows them to push mean unit activations closer to zero like batch normalization.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

Comment (thomelane) [reply python/guide/module/gluon/activation.md]: which may allows -> which may allow

File: python/guide/module/gluon/activation.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+```python
+visualize_activation(data, gluon.nn.ELU(), "ELU")
+```
+
+
+![png](img/output_26_0.png)
+
+
+### SELU
+SELU stands for Scaled Exponential Linear Unit and was introduced by [Klambuer et al](https://arxiv.org/abs/1706.02515) and is a modification of the ELU that improves the normalization of it's outputs towards a zero mean and unit variance.
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: it's -> its

File: python/guide/module/gluon/activation.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+If you are looking to answer the question, 'which activation function should I use for my neural network model?', you should probably go with *ReLU*. Unless you're trying to implement something like a gating mechanism, like in LSTMs or GRU cells, then you should opt for sigmoid and/or tanh in those cells. However, if you have a working model architecture and you're trying to improve it's performance by swapping out activation functions or treating the activation function as a hyperparameter, then you may want to try hand-designed activations like SELU or a function discovered by reinforcement learning and exhaustive search like Swish. This guide describes these activation functions and others implemented in MXNet in detail.
+
+### Visualizing Activations
+In order to compare the various activation functions and to understand the nuances of their differences we have a snippet of code to plot the activation functions (used in the forward pass) and their gradients (used in the backward pass).
+
+
+```python
+import numpy as np
+import mxnet
+from mxnet import nd, gluon, autograd
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: Just `import mxnet as mx` and use full path to activation function.

File: python/guide/module/gluon/activation.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+from matplotlib import pyplot as plt
+%matplotlib inline
+
+data = np.linspace(-10, 10, 501)
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: Would just put this into the function to avoid having to keep passing it through.

File: python/guide/module/gluon/activation.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+def visualize_activation(data, activation_fn, activation):
```
Comment (thomelane) [inline python/guide/module/gluon/activation.md]: And in addition to removing data, can we infer activation name from activation_fn? Would only need single arg then.

File: python/guide/module/gluon/activation.md (hunk 13 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: will address thanks

File: python/guide/module/gluon/activation.md (hunk 3 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: thanks for catching

File: python/guide/module/gluon/activation.md (hunk 4 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: addressed.

File: python/guide/module/gluon/activation.md (hunk 5 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: addressed

File: python/guide/module/gluon/activation.md (hunk 6 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: will change

File: python/guide/module/gluon/activation.md (hunk 7 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: again thanks

File: python/guide/module/gluon/activation.md (hunk 8 above)
Comment (sad-) [reply python/guide/module/gluon/activation.md]: addressed

**Comment by thomelane** [comment]
Closing out. All got merged into master, even though it doesn't show up correctly on here (https://github.com/mli/new-docs/commit/9aeae7dabd8ce1a11d1f90d6e22546ff08ef4b6b and https://github.com/mli/new-docs/commit/7ad95681f4113ccca230076bda7c18c421c42c13). Many thanks again Soji!
